crypto/s2n_locking.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <openssl/crypto.h>
+#include <pthread.h>
+
+#include "crypto/s2n_locking.h"
+#include "crypto/s2n_openssl.h"
+#include "utils/s2n_mem.h"
+#include "utils/s2n_safety.h"
+
+/* Writing multithreaded applications using Openssl-1.0.2
+ * requires calling CRYPTO_set_locking_callback.
+ * If the callback is not set, locks are no-ops and unexpected
+ * behavior may occur, particularly for RSA and X509.
+ *
+ * In the past s2n-tls relied on customers setting the callback
+ * themselves, but that seems unnecessary since other parts of
+ * the library (like fork detection) already rely on the pthreads library.
+ *
+ * For more information:
+ * https://www.openssl.org/blog/blog/2017/02/21/threads/
+ * https://www.openssl.org/docs/man1.0.2/man3/threads.html
+ */
+
+#define S2N_MUTEXES(mem) ((pthread_mutex_t *) (void*) (mem).data)
```

Comment:
We are synchronizing on reads. We have no logic to treat reads and writes differently. We only check CRYPTO_LOCK.

I think looking at the newer locking API definitions gives a better idea of what the modes mean:
> #define CRYPTO_w_lock(type)    \
        CRYPTO_lock(CRYPTO_LOCK|CRYPTO_WRITE,type,\_\_FILE\_\_,\_\_LINE\_\_)
 #define CRYPTO_w_unlock(type)  \
        CRYPTO_lock(CRYPTO_UNLOCK|CRYPTO_WRITE,type,\_\_FILE\_\_,\_\_LINE\_\_)
 #define CRYPTO_r_lock(type)    \
        CRYPTO_lock(CRYPTO_LOCK|CRYPTO_READ,type,\_\_FILE\_\_,\_\_LINE\_\_)
 #define CRYPTO_r_unlock(type)  \
        CRYPTO_lock(CRYPTO_UNLOCK|CRYPTO_READ,type,\_\_FILE\_\_,\_\_LINE\_\_)

Reply:
Sorry, I mean can't we get away with read-write locks instead of a mutex? The latter blocks parallel reading - but is that necessary in our case?

I'm not sure what OpenSSL requires and they do use mutexes in the [example implementation](https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/threads/th-lock.c).

There is also a performance aspect, of course. Sometimes, mutex could be more performant and I don't really have a good intuition.

Reply:
Ok - from offline chat

Reply:
From the offline chat: the example implementation and the known customer implementations all just use a mutex. Implementing read/write locking isn't currently an ask, so is a potential risk with no gain.